pkg/messageprocessors/javascript/javascript.go

```diff
@@ -205,21 +278,72 @@ func (h *host) decodeUplink(ctx context.Context, ids *ttnpb.EndDeviceIdentifiers
 		return err
 	}
 
-	var output decodeUplinkOutput
+	var output uplinkDecoderOutput
 	err = valueAs(&output)
 	if err != nil {
 		return errOutput.WithCause(err)
 	}
-	if len(output.Errors) > 0 {
-		return errOutputErrors.WithAttributes("errors", strings.Join(output.Errors, ", "))
-	}
 
-	s, err := gogoproto.Struct(output.Data)
+	if errs := output.Decoded.Errors; len(errs) > 0 {
+		return errOutputErrors.WithAttributes("errors", strings.Join(errs, ", "))
+	}
+	decodedPayload, err := gogoproto.Struct(output.Decoded.Data)
 	if err != nil {
 		return errOutput.WithCause(err)
 	}
-	msg.DecodedPayload = s
-	msg.DecodedPayloadWarnings = output.Warnings
+	msg.DecodedPayload, msg.DecodedPayloadWarnings = decodedPayload, output.Decoded.Warnings
+
+	if normalized := output.Normalized; normalized != nil {
+		if errs := normalized.Errors; len(errs) > 0 {
+			return errOutputErrors.WithAttributes("errors", strings.Join(errs, ", "))
+		}
+		if normalized.Data == nil {
+			return errOutput.New()
+		}
+		// The returned data can be an array of measurements or a single measurement object.
+		var ms []map[string]interface{}
+		if val := reflect.ValueOf(normalized.Data); val.Kind() == reflect.Slice {
+			ms = make([]map[string]interface{}, val.Len())
+			for i := 0; i < val.Len(); i++ {
+				measurement, ok := val.Index(i).Interface().(map[string]interface{})
+				if !ok {
+					return errOutput.New()
+				}
+				ms[i] = measurement
+			}
+		} else {
+			measurement, ok := normalized.Data.(map[string]interface{})
+			if !ok {
+				return errOutput.New()
+			}
+			ms = []map[string]interface{}{measurement}
+		}
+		normalizedPayload := make([]*pbtypes.Struct, len(ms))
+		for i := range ms {
+			pb, err := gogoproto.Struct(ms[i])
+			if err != nil {
+				return errOutput.WithCause(err)
+			}
+			normalizedPayload[i] = pb
+		}
+		// Validate the normalized payload.
+		_, err := normalizedpayload.Parse(normalizedPayload)
+		if err != nil {
+			return errOutput.WithCause(err)
+		}
+		msg.NormalizedPayload, msg.NormalizedPayloadWarnings = normalizedPayload, normalized.Warnings
+	} else {
+		// If the normalizer is not set, the decoder may return already normalized payload.
+		// This is a best effort attempt to parse the decoded payload as normalized payload.
+		// If that does not return an error, the decoded payload is assumed to be normalized.
+		normalizedPayload := []*pbtypes.Struct{
+			decodedPayload,
+		}
+		_, err := normalizedpayload.Parse(normalizedPayload)
+		if err == nil {
+			msg.NormalizedPayload, msg.NormalizedPayloadWarnings = normalizedPayload, nil
+		}
```

Comment:
@adriansmares you expressed concerns with a performance hit of parsing each and every decoded uplink through the normalized payload validator (https://github.com/TheThingsNetwork/lorawan-stack/issues/5429#issuecomment-1196598347) (1/3)